src/TinyMvvm.Forms/NavigationHelper/FormsNavigationHelper.cs

```diff
@@ -0,0 +1,440 @@
+
+                        if (prevPage.BindingContext is ViewModelBase tviewModel)
+                        {
+                            tviewModel.ReturningParameter = parameter;
+                            tviewModel.ReturningHasRun = true;
+
+                            Device.BeginInvokeOnMainThread(async () => await tviewModel.Returning());
+                        }
+
+                        var tpage = await selected.Navigation.PopAsync();
+
+                       
+
+                        return;
+                    }
+                }
+                else if (Application.Current.MainPage is MasterDetailPage masterDetailPage)
+                {
+                    if (masterDetailPage.Detail is TabbedPage tabbedPage)
+                    {
+                        if (tabbedPage.CurrentPage.Navigation != null)
+                        {
+                            var prevPage = tabbedPage.CurrentPage.Navigation.NavigationStack[0];
+
+                            if (prevPage.BindingContext is ViewModelBase tviewModel)
+                            {
+                                tviewModel.ReturningParameter = parameter;
+                                tviewModel.ReturningHasRun = true;
+
+                                Device.BeginInvokeOnMainThread(async () => await tviewModel.Returning());
+                            }
+
+                           await tabbedPage.CurrentPage.Navigation.PopAsync();
+
+                            
+                            return;
+                        }
+                    }
+                    if (masterDetailPage.Detail.Navigation != null)
+                    {
+                        var prevPage = masterDetailPage.Detail.Navigation.NavigationStack[0];
+
+                        if (prevPage.BindingContext is ViewModelBase mdviewModel)
+                        {
+                            mdviewModel.ReturningParameter = parameter;
+                            mdviewModel.ReturningHasRun = true;
+
+                            Device.BeginInvokeOnMainThread(async () => await mdviewModel.Returning());
+                        }
+
+                        await masterDetailPage.Detail.Navigation.PopAsync();
+
+                        
+
+                        return;
+                    }
+                }
+
+                var prevPage1 = Application.Current.MainPage.Navigation.NavigationStack[0];
+
+                if (prevPage1.BindingContext is ViewModelBase viewModel)
+                {
+                    viewModel.ReturningParameter = parameter;
+                    viewModel.ReturningHasRun = true;
+
+                    Device.BeginInvokeOnMainThread(async () => await viewModel.Returning());
+                }
+
+                await Application.Current.MainPage.Navigation.PopAsync();
+
+                
+            }
+            else
+            {
+                var prevPage = Application.Current.MainPage.Navigation.NavigationStack[0];
```

Comment:
This might need to be `ModalStack`. In my app, I have the following:

- TabbedPage
   - ContentPage w/ Navigation
      - Opens Modal ContentPage w/ Navigation
         - Opens ContentPage

In this case, when pressing back on the bottom level page, the previous page only exists in `ModalStack` (`NavigationStack` is empty).

Reply:
This is probably not 100% correct. The `ModalStack[0]` is the `NavigationPage` which has a `CurrentPage` of where we're coming from and a `Pages` `IEnumerable` of all the pages before it. This might be a bit more work than just replacing this with `ModalStack` :D

![image](https://user-images.githubusercontent.com/2419439/101138676-cfe1f200-3610-11eb-9332-cc2b499bd426.png)